proxy/src/main/java/com/wavefront/agent/QueuedAgentService.java

```diff
@@ -451,15 +454,18 @@ public Response postPushData(UUID agentId, UUID workUnitId, Long currentMillis,
   }
 
   private void addTaskToSmallestQueue(ResubmissionTask taskToRetry) {
-    TaskQueue<ResubmissionTask> queue = getSmallestQueue();
+    ResubmissionTaskQueue queue = getSmallestQueue();
     if (queue != null) {
-      synchronized (queue) {
+      try {
```

Comment:
on second thought, i guess putting it outside the loop can also lock without releasing (depending on how the compiler weaves the code together), anyhow, it's just convention to lock then try :)